File: src/learning_gem5/tommy/time_example_object.cc
#include "learning_gem5/tommy/time_example_object.hh"

#include <iostream>

#include "base/trace.hh"
#include "debug/TimeExampleObject.hh"

namespace gem5
{

TimeExampleObject::TimeExampleObject(const TimeExampleObjectParams &params)
    : SimObject(params),
      event([this] { processEvent(); }, name()),
      latency(params.latency),
      timesLeft(params.times_lelft)
{ DPRINTF(TimeExampleObject, "Hello World! From a SimObject!\n"); }

void
TimeExampleObject::processEvent()
{
    timesLeft--;
    DPRINTF(TimeExampleObject, "Processing the event! %d left\n", timesLeft);

    if (timesLeft <= 0) {
        DPRINTF(TimeExampleObject, "Last event !\n");
    } else {
        schedule(event, curTick() + latency);
    }
}

void
TimeExampleObject::startup()
{ schedule(event, 100); }

} // namespace gem5

File: src/learning_gem5/tommy/TimeExampleObject.py
from m5.params import *
from m5.SimObject import SimObject


class TimeExampleObject(SimObject):
    type = "TimeExampleObject"
    cxx_header = "learning_gem5/tommy/time_example_object.hh"
    cxx_class = "gem5::TimeExampleObject"
    latency = Param.Latency("1ns", "latency between events")
    times_lelft = Param.Int(10, "how many times we want to trigger the events")
